src/PrestaShopBundle/Form/Admin/Sell/CartRule/EventListener/DiscountListener.php

```diff
@@ -0,0 +1,101 @@
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Form\Admin\Sell\CartRule\EventListener;
+
+use PrestaShop\PrestaShop\Core\Domain\ValueObject\Reduction;
+use PrestaShop\PrestaShop\Core\Form\ChoiceProvider\DiscountApplicationChoiceProvider;
+use PrestaShopBundle\Form\FormCloner;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+
+class DiscountListener implements EventSubscriberInterface
+{
+    /**
+     * @var DiscountApplicationChoiceProvider
+     */
+    private $discountApplicationChoiceProvider;
+
+    /**
+     * @var FormCloner
+     */
+    private $formCloner;
+
+    public function __construct(
+        DiscountApplicationChoiceProvider $discountApplicationChoiceProvider,
+        FormCloner $formCloner
+    ) {
+        $this->discountApplicationChoiceProvider = $discountApplicationChoiceProvider;
+        $this->formCloner = $formCloner;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            FormEvents::PRE_SET_DATA => 'adaptDiscountChoices',
+            FormEvents::PRE_SUBMIT => 'adaptDiscountChoices',
+        ];
+    }
+
+    /**
+     * @param FormEvent $event
+     */
+    public function adaptDiscountChoices(FormEvent $event): void
+    {
+        $data = $event->getData();
+        if (!isset($data['reduction']['type'])) {
+            return;
+        }
+
+        $form = $event->getForm();
+        $discountApplicationField = $form->get('discount_application');
+        // adjust discount application choices depending on reduction type data
+        $choices = $this->getChoicesByType($data['reduction']['type']);
+        $newDiscountApplicationField = $this->formCloner->cloneForm($discountApplicationField, [
+            'choices' => $choices,
+            'attr' => [
+                // these attributes are needed for js to change choices when reduction type changes
+                'data-amount-choices' => json_encode($this->getChoicesByType(Reduction::TYPE_AMOUNT)),
+                'data-percentage-choices' => json_encode($this->getChoicesByType(Reduction::TYPE_PERCENTAGE)),
+            ],
```

Comment:
This could be moved into the form type as it's always the same and uses hard-coded types